service/history/workflow/mutable_state_impl.go

```diff
@@ -4955,33 +4962,78 @@ func (ms *MutableStateImpl) updateActivityInfoForRetries(
 	nextAttempt int32,
 	activityFailure *failurepb.Failure,
 ) {
-	ms.updateActivity(ai, func(info *persistencespb.ActivityInfo, impl *MutableStateImpl) *persistencespb.ActivityInfo {
-		ai = UpdateActivityInfoForRetries(
-			ai,
-			ms.GetCurrentVersion(),
-			nextAttempt,
-			ms.truncateRetryableActivityFailure(activityFailure),
-			timestamppb.New(nextScheduledTime),
-		)
-		return ai
+	ms.UpdateActivityWithCallback(ai, func(info *persistencespb.ActivityInfo, mutableState MutableState) {
+		mutableStateImpl, ok := mutableState.(*MutableStateImpl)
+		if ok {
+			ai = UpdateActivityInfoForRetries(
+				ai,
+				mutableStateImpl.GetCurrentVersion(),
+				nextAttempt,
+				mutableStateImpl.truncateRetryableActivityFailure(activityFailure),
+				timestamppb.New(nextScheduledTime),
+			)
+		}
 	})
 }
 
-func (ms *MutableStateImpl) updateActivity(ai *persistencespb.ActivityInfo, updateCallback func(*persistencespb.ActivityInfo, *MutableStateImpl) *persistencespb.ActivityInfo) {
-	// we need to store activity info size since pendingActivityInfoIDs holds pointers to activity
-	// info and if prev found it points to the same activity info as ai, so updating ai will cause
-	// size of prev change.
+func (ms *MutableStateImpl) UpdateActivityWithCallback(
+	ai *persistencespb.ActivityInfo,
+	updateCallback UpdateActivityCallback,
+) {
+	// incoming activity info can be a pointer to the activity info in pendingActivityInfoIDs
+	// we need to store activity info size since pendingActivityInfoIDs holds pointers to activity info
+	// if prev found it can point to the same activity info as incoming activity info,
 	var originalSize int
 	if prev, ok := ms.pendingActivityInfoIDs[ai.ScheduledEventId]; ok {
 		originalSize = prev.Size()
 	}
+
 	updateCallback(ai, ms)
 
 	ms.approximateSize += ai.Size() - originalSize
 	ms.updateActivityInfos[ai.ScheduledEventId] = ai
 	ms.syncActivityTasks[ai.ScheduledEventId] = struct{}{}
 }
 
+func (ms *MutableStateImpl) UpdatePausedEntitiesSearchAttribute() error {
```

Comment:
This should be called after applying replication tasks too.

Reply:
this I didn't understand. Why? This will add _another_ search attribute? Or we don't replicate visibility?

Reply:
We don't replicate visibility tasks.

Reply:
ok, then it needs to be re-applied.
I update UpdateActivityInfo (it is called after replication to sync activities)